In [2]:
import pandas as pd

dados = pd.read_csv("dados_tratados.csv")

dados.head()

customerID  Churn customer_gender  customer_SeniorCitizen customer_Partner  \
0  0002-ORFBO      0          Female                       0              Yes   
1  0003-MKNFE      0            Male                       0               No   
2  0004-TLHLJ      1            Male                       0               No   
3  0011-IGKFF      1            Male                       1              Yes   
4  0013-EXCHZ      1          Female                       1              Yes   

  customer_Dependents  customer_tenure phone_PhoneService phone_MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet_InternetService  ... internet_OnlineBackup  \
0                      DSL  ...                   Yes   
1                      DSL  ...                    No   
2              Fiber optic  ...                    No   
3              Fiber optic  ...                   Yes   
4              Fiber optic  ...                    No   

  internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
0                        No                  Yes                  Yes   
1                        No                   No                   No   
2                       Yes                   No                   No   
3                       Yes                   No                  Yes   
4                        No                  Yes                  Yes   

  internet_StreamingMovies account_Contract account_PaperlessBilling  \
0                       No         One year                      Yes   
1                      Yes   Month-to-month                       No   
2                       No   Month-to-month                      Yes   
3                      Yes   Month-to-month                      Yes   
4                       No   Month-to-month                      Yes   

  account_PaymentMethod account_Charges_Monthly  account_Charges_Total  
0          Mailed check                    65.6                 593.30  
1          Mailed check                    59.9                 542.40  
2      Electronic check                    73.9                 280.85  
3      Electronic check                    98.0                1237.85  
4          Mailed check                    83.9                 267.40  

[5 rows x 21 columns]

In [3]:
dados_processados = dados.drop('customerID', axis=1)

dados_processados = pd.get_dummies(dados_processados, drop_first=True)

print("Dimensões do DataFrame após o pré-processamento:", dados_processados.shape)
display(dados_processados.head())

Dimensões do DataFrame após o pré-processamento: (7043, 31)


Churn  customer_SeniorCitizen  customer_tenure  account_Charges_Monthly  \
0      0                       0                9                     65.6   
1      0                       0                9                     59.9   
2      1                       0                4                     73.9   
3      1                       1               13                     98.0   
4      1                       1                3                     83.9   

   account_Charges_Total  customer_gender_Male  customer_Partner_Yes  \
0                 593.30                 False                  True   
1                 542.40                  True                 False   
2                 280.85                  True                 False   
3                1237.85                  True                  True   
4                 267.40                 False                  True   

   customer_Dependents_Yes  phone_PhoneService_Yes  \
0                     True                    True   
1                    False                    True   
2                    False                    True   
3                    False                    True   
4                    False                    True   

   phone_MultipleLines_No phone service  ...  \
0                                 False  ...   
1                                 False  ...   
2                                 False  ...   
3                                 False  ...   
4                                 False  ...   

   internet_StreamingTV_No internet service  internet_StreamingTV_Yes  \
0                                     False                      True   
1                                     False                     False   
2                                     False                     False   
3                                     False                      True   
4                                     False                      True   

   internet_StreamingMovies_No internet service  internet_StreamingMovies_Yes  \
0                                         False                         False   
1                                         False                          True   
2                                         False                         False   
3                                         False                          True   
4                                         False                         False   

   account_Contract_One year  account_Contract_Two year  \
0                       True                      False   
1                      False                      False   
2                      False                      False   
3                      False                      False   
4                      False                      False   

   account_PaperlessBilling_Yes  \
0                          True   
1                         False   
2                          True   
3                          True   
4                          True   

   account_PaymentMethod_Credit card (automatic)  \
0                                          False   
1                                          False   
2                                          False   
3                                          False   
4                                          False   

   account_PaymentMethod_Electronic check  account_PaymentMethod_Mailed check  
0                                   False                                True  
1                                   False                                True  
2                                    True                               False  
3                                    True                               False  
4                                   False                                True  

[5 rows x 31 columns]

DataFrame dados_processados agora está no formato para ser usado por algoritmos de Machine Learning: totalmente numérico.

Passamos de 21 colunas para 31, pois as variáveis de texto foram transformadas em colunas binárias.

**Separação dos Dados em Treino e Teste**

DataFrame em duas partes:
X: Todas as colunas que serão usadas como "pistas" para a previsão (as features).

y: A coluna que queremos prever (Churn).

In [4]:
from sklearn.model_selection import train_test_split

# Separar as features (X) e o alvo (y)
X = dados_processados.drop('Churn', axis=1)
y = dados_processados['Churn']

# Dividir os dados em conjuntos de treino e teste

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# VERIFICAÇÃO: Vamos verificar as dimensões dos conjuntos criados
print("Dimensões de X_train:", X_train.shape)
print("Dimensões de X_test:", X_test.shape)
print("Dimensões de y_train:", y_train.shape)
print("Dimensões de y_test:", y_test.shape)

Dimensões de X_train: (4930, 30)
Dimensões de X_test: (2113, 30)
Dimensões de y_train: (4930,)
Dimensões de y_test: (2113,)


Dimensões de X_train: (4930, 30): 4930 clientes para treinar modelo, e para cada um, temos 30 características (colunas) para usar como "pistas".

Dimensões de X_test: (2113, 30): 2113 clientes para o teste final. Note que o número de colunas (30) é o mesmo, o que é crucial.

Dimensões de y_train: (4930,): Para cada um dos 4930 clientes de treino, temos a resposta correta (se ele evadiu ou não).

Dimensões de y_test: (2113,): E para os 2113 clientes de teste, também tem o gabarito, para comparar com as previsões do modelo.

A soma 4930 + 2113 = 7043, que é o número total de clientes.

A divisão 70/30 foi feita perfeitamente.

In [5]:
from sklearn.linear_model import LogisticRegression

modelo = LogisticRegression(max_iter=1000)

modelo.fit(X_train, y_train)

print("Modelo treinado com sucesso!")

Modelo treinado com sucesso!


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


**Todas as variáveis na mesma escala**

A técnica mais comum é o StandardScaler, que transforma cada coluna para que ela tenha média 0 e desvio padrão 1.

**inserir** o escalonamento no fluxo, é um passo de pré-processamento que deve vir depois da divisão treino/teste.

In [6]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)

X_test_scaled = scaler.transform(X_test)

modelo = LogisticRegression()

modelo.fit(X_train_scaled, y_train)

print("Modelo treinado com sucesso e sem avisos!")

Modelo treinado com sucesso e sem avisos!


Para avaliar a performance, usarei três métricas importantes:

Acurácia (Accuracy): A porcentagem de previsões que o modelo acertou no total. É a métrica mais simples.

Matriz de Confusão (Confusion Matrix): Uma tabela que nos mostra em detalhes os tipos de acertos e erros do modelo (ex: previu que o cliente sairia e ele realmente saiu).

Relatório de Classificação (Classification Report): Fornece métricas mais detalhadas como precisão e recall, que são muito úteis para entender a performance do modelo para cada classe (Churn "Sim" e Churn "Não").

In [7]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

y_pred = modelo.predict(X_test_scaled)

# Calcular a acurácia
acuracia = accuracy_score(y_test, y_pred)
print(f"Acurácia do modelo: {acuracia * 100:.2f}%")

print("\n--------------------------------------------------------\n")

print("Matriz de Confusão:")
print(confusion_matrix(y_test, y_pred))

print("\n--------------------------------------------------------\n")

# Exibir o Relatório de Classificação
print("Relatório de Classificação:")
print(classification_report(y_test, y_pred, target_names=['Não Churn (0)', 'Churn (1)']))

Acurácia do modelo: 81.50%

--------------------------------------------------------

Matriz de Confusão:
[[1410  152]
 [ 239  312]]

--------------------------------------------------------

Relatório de Classificação:
               precision    recall  f1-score   support

Não Churn (0)       0.86      0.90      0.88      1562
    Churn (1)       0.67      0.57      0.61       551

     accuracy                           0.81      2113
    macro avg       0.76      0.73      0.75      2113
 weighted avg       0.81      0.81      0.81      2113



**Relatório de Classificação**

O relatório transforma os números da matriz de confusão em métricas percentuais, que são mais fáceis de comparar.
Focando na linha mais importante para o negócio, que é a Churn (1):

Precision (0.67): De todos os clientes que o modelo alertou como "potencial Churn", 67% deles realmente cancelaram. Mostra que os "alarmes" do modelo são razoavelmente confiáveis.

Recall (0.57): métrica mais crítica. De todos os clientes que realmente cancelaram, o modelo conseguiu identificar 57% deles.

F1-score (0.61): média harmônica entre a precisão e o recall. Serve como uma métrica única que equilibra os dois.

**Conclusão e PróximosPassos**

O primeiro modelo de Regressão Logística é bom em prever os clientes que vão ficar (recall de 90%), mas tem uma performance moderada para prever os que vão sair (recall de 57%).

Ele consegue identificar mais da metade dos clientes que estão em risco, o que já é extremamente valioso para a empresa.

Onde podem melhorar? O principal ponto é aumentar o recall para a classe "Churn (1)". Queremos "deixar passar" o menor número possível de clientes que estão prestes a cancelar, mesmo que isso signifique aumentar um pouco o número de falsos positivos.